darwin.c

```diff
@@ -2,37 +2,55 @@
 #import <Foundation/Foundation.h>
 #import <SystemConfiguration/SCPreferences.h>
 #import <SystemConfiguration/SCNetworkConfiguration.h>
+
 #include <sys/syslimits.h>
 #include <sys/stat.h>
 #include <mach-o/dyld.h>
 #include "common.h"
 
-int setUid()
+/* === implement details === */
+
+typedef Boolean (*visitor) (SCNetworkProtocolRef proxyProtocolRef, NSDictionary* oldPreferences, bool turnOn, const char* pacUrl);
+
+Boolean showAction(SCNetworkProtocolRef proxyProtocolRef/*unused*/, NSDictionary* oldPreferences, bool turnOn/*unused*/, const char* pacUrl/*unused*/)
 {
-  char exeFullPath [PATH_MAX];
-  uint32_t size = PATH_MAX;
-  if (_NSGetExecutablePath(exeFullPath, &size) != 0)
-  {
-    printf("Path longer than %d, should not occur!!!!!", size);
-    return SYSCALL_FAILED;
-  }
-  if (chown(exeFullPath, 0, 0) != 0) // root:wheel
-  {
-    puts("Error chown");
-    return NO_PERMISSION;
-  }
-  if (chmod(exeFullPath, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH | S_ISUID) != 0)
-  {
-    puts("Error chmod");
-    return NO_PERMISSION;
+  NSNumber* on = [oldPreferences valueForKey:(NSString*)kSCPropNetProxiesProxyAutoConfigEnable];
+  NSString* nsOldPacUrl = [oldPreferences valueForKey:(NSString*)kSCPropNetProxiesProxyAutoConfigURLString];
+  if ([on intValue] == 1) {
+    printf("%s\n", [nsOldPacUrl UTF8String]);
   }
-  return RET_NO_ERROR;
+  return TRUE;
 }
 
-int togglePac(bool turnOn, const char* pacUrl)
+Boolean toggleAction(SCNetworkProtocolRef proxyProtocolRef, NSDictionary* oldPreferences, bool turnOn, const char* pacUrl)
 {
   NSString* nsPacUrl = [[NSString alloc] initWithCString: pacUrl encoding:NSUTF8StringEncoding];
   NSString* nsOldPacUrl;
+  NSMutableDictionary *newPreferences = [NSMutableDictionary dictionaryWithDictionary: oldPreferences];
+  Boolean success;
+
+  if(turnOn == true) {
+    [newPreferences setValue:[NSNumber numberWithInt:1] forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigEnable];
+    [newPreferences setValue:nsPacUrl forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigURLString];
+  } else {
+    nsOldPacUrl = [oldPreferences valueForKey:(NSString*)kSCPropNetProxiesProxyAutoConfigURLString];
+    // we turn pac off only if the option is set and pac url has the provided
+    // prefix.
+    if (nsPacUrl.length == 0 || [nsOldPacUrl hasPrefix:nsPacUrl]) {
+      [newPreferences setValue:[NSNumber numberWithInt:0] forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigEnable];
+      [newPreferences setValue:@"" forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigURLString];
+    }
+  }
+
+  success = SCNetworkProtocolSetConfiguration(proxyProtocolRef, (__bridge CFDictionaryRef)newPreferences);
+  if(!success) {
+    NSLog(@"Failed to set Protocol Configuration");
+  }
+  return success;
```

Comment:
Oh, I understand now. This boolean gets translated into an error before returning from `show()` so it's okay.